### PR TITLE
Create backend api endpoints

### PR DESCRIPTION
**Pull request by markswinimer**
made api endpoints in back end. Added search function with react.js frontend. Made enzyme tests for new react components and containers, but no tests for fetch API (we were advised not to test these features).

### CODE REVIEW

**Comment by laurado** [comment]
Make sure that in the future, you only work on one feature per feature branch! For example, this is the create-backend-api-endpoints branch, so make a new branch for your react components and related tests.

**Comment by laurado** [comment]
Create tests for your API endpoints!

File: app/controllers/api/v1/restaurants_controller.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
+class Api::V1::RestaurantsController < ApplicationController
+
+  def index
+    render json: { restaurant: Restaurant.all.sample }
+  end
+
+  def create
+    body = JSON.parse(request.body.read)
```
Comment (enthusiastick) [inline app/controllers/api/v1/restaurants_controller.rb:8]: In an ideal world, you shouldn't have to `JSON.parse()` the request body in an API controller action this way. The Rails `params` magic should handle that for you, assuming that everything has been correctly formulated.

File: app/controllers/api/v1/restaurants_controller.rb [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      render json: { restaurant: restaurant }
+    end
+  end
+
+  private
+
+  def restaurant_params(body)
+    body.require(:restaurant).permit(:name, :address, :city, :state, :zip, :dollar_value,
+      :description, :picture, :phone_number, :website_url, :user)
+  end
```
Comment (enthusiastick) [inline app/controllers/api/v1/restaurants_controller.rb:20]: I love seeing `strong_params`, but I'm not sure you're ever actually utilizing them in this controller -- you appear to just be passing the entire unfiltered hash assigned to `body` (on line 8) to the `.new()` method. 😞

File: app/controllers/api/v1/searches_controller.rb [hunk 3]
```diff
@@ -0,0 +1,9 @@
+class Api::V1::SearchesController < ApplicationController
+
+  def create
+    body = JSON.parse(request.body.read)
+    restaurant = Restaurant.where("name LIKE ? OR city LIKE ?", "%#{body["value"]}%", "%#{body["value"]}%" )
```
Comment (enthusiastick) [inline app/controllers/api/v1/searches_controller.rb:5]: Nice use of that SQL `LIKE` for your search. 👍

File: app/javascript/react/src/containers/SearchResultsContainer.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+          city={restaurant.city}
+          state={restaurant.state}
+          zip={restaurant.zip}
+          dollar_value={restaurant.dollar_value}
+          description={restaurant.description}
+          picture={restaurant.picture}
+          phone_number={restaurant.phone_number}
+          website_url={restaurant.website_url}
+          dining_type={restaurant.restaurant_dining_type}
+          food_type={restaurant.restaurant_food_type}
```
Comment (enthusiastick) [inline app/javascript/react/src/containers/SearchResultsContainer.js:27]: These props are a little long for my taste. Why not simply pass down `restaurant={restaurant}`, and then decompose in the presentational component instead?

File: app/javascript/react/test/containers/RestaurantFormContainerTest.spec.js [hunk 5]
```diff
@@ -46,7 +46,7 @@ describe('RestaurantFormContainer', () => {
   });
 
   it('should submit the form on click', () => {
-    console.log(wrapper.find('form').props().onSubmit())
+    wrapper.find('form').props().onSubmit()
     expect(RestaurantFormContainer.prototype.handleSubmit).toHaveBeenCalled();
   })
```
Comment (enthusiastick) [inline app/javascript/react/test/containers/RestaurantFormContainerTest.spec.js:52]: Nice. 😃

File: app/models/restaurant.rb [hunk 6]
```diff
@@ -1,4 +1,11 @@
 class Restaurant < ApplicationRecord
+
+
+  def self.search(search)
+    where("name ilike ?", "%#{search}%")
+    # where(dollar_value: search)
```
Comment (enthusiastick) [inline app/models/restaurant.rb:6]: I am not wholly sold on this as a pattern (esp. with the commented-out code just beneath it, which as a best practice should not be merged into `master`.)

File: app/controllers/api/v1/restaurants_controller.rb (hunk 2 above)
Comment (markswinimer) [reply app/controllers/api/v1/restaurants_controller.rb:20]: Hi Eben, we worked with Brianna for a few to try and pass the request body into the strong params but were unsuccessful. I'm not sure we had the correct syntax. How can we pass the request body into the strong params?

File: spec/controllers/api/v1/restaurants_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+      expect(response.content_type).to eq("application/json")
+
+      expect(returned_json).to be_kind_of(Hash)
+      expect(returned_json).to_not be_kind_of(Array)
+      expect(returned_json["name"]).to eq "Panera"
+
+
+    end
+  end
+end
```
Comment (enthusiastick) [inline spec/controllers/api/v1/restaurants_controller_spec.rb:53]: I love API tests! This one could be cleaned up a little (especially if you used `FactoryGirl`) but I am happy to see it either way.

File: spec/controllers/api/v1/searches_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+      returned_json = JSON.parse(response.body)["restaurant"]
+      expect(response.status).to eq 200
+      expect(response.content_type).to eq("application/json")
+
+      expect(returned_json).to be_kind_of(Array)
+      expect(returned_json.length).to eq 1
+      expect(returned_json[0]["city"]).to eq "Boston"
+    end
+  end
+end
```
Comment (enthusiastick) [inline spec/controllers/api/v1/searches_controller_spec.rb:39]: ❤️

**Comment by enthusiastick** [comment]
I'll allow it

File: app/controllers/api/v1/restaurants_controller.rb (hunk 2 above)
Comment (enthusiastick) [reply app/controllers/api/v1/restaurants_controller.rb:20]: You would probably have to transform the top-level key from `body` to `restaurant`. Which you could do inside the controller, given your current implementation.

**Pull request merged**